Segfault from FFI closure function in interpreter

There is a bug causing a segmentation fault when using FFI closured function pointers in the interpreter.

A triggering example is `XML.parse("<foo>")`. The setup for `libxml2` configures some function pointers for memory management via [`LibXML.xmlGcMemSetup`](https://github.com/crystal-lang/crystal/blob/37497c9ef2bbfd552e53f1a8463f70a421c4b50d/src/xml/libxml2.cr#L317-L328).
These function pointers are then used from within libxml2 when it needs to allocate memory, for example. This happens when parsing an XML document. Apparently, the closured function that's supposed to be called from the closure wrapper is broken: The `user_data` pointer (which holds a reference to the interpreter and context instances) points to a very low address like `0x43` (c.f. [`Context.ffi_closure_fun`](https://github.com/crystal-lang/crystal/blob/37497c9ef2bbfd552e53f1a8463f70a421c4b50d/src/compiler/crystal/interpreter/context.cr#L266)).

I have been able to reproduce what looks like the same bug without libxml:
```c
// libclosure.c
// build: gcc -shared -fPIC .test/lib_closure.c -o libclosure.so
int (*foo_ptr)(int);

void set_foo(int (*f)(int)) {
  foo_ptr = f;
}
```

```cr
#! bin/crystal i
@[Link("./libclosure.so")]
lib LibClosure
  fun set_foo(fun_ptr : LibC::Int -> LibC::Int) : Void

  $foo_ptr : LibC::Int -> LibC::Int
end

LibClosure.set_foo -> (i : LibC::Int) { i * 100 }
LibClosure.foo_ptr.pointer # => Pointer(Void)@0x7f1ad3d98f40
LibClosure.foo_ptr.call(4) # Error: Invalid memory access (signal 11) at address 0x0
```

It seems to be working fine when the function pointer in a global variable is called from another C function like this:
```c
int call_foo_ptr() {
  (*foo_ptr)(42);
}
```

When the value of the global variable is assigned from Crystal, it can be called from Crystal. But now it errors when called from C:
```cr
LibClosure.foo_ptr = -> (i : LibC::Int) { i * 100 }
LibClosure.foo_ptr.call(4) # => 400
LibClosure.foo_ptr.pointer # => Pointer(Void)@0x7fd71f787180
LibClosure.call_foo_ptr # Error: Invalid memory access (signal 11) at address 0x7fd71f787180
```
Note that here the address of the function pointer triggers the invalid access.

Oh, it's because the C file is wrong. Don't you get this when you compile the lib?

```
spec/compiler/data/libs/funptr.c:13:1: warning: non-void function does not return a value [-Wreturn-type]
```

A `return` is missing in `funptr_call`:

```crystal
int funptr_call() {
  return (*fun_ptr)(42); // In your branch this return is missing
}
```

Maybe your C compiler adds a return 🤷

--- a/src/compiler/crystal/ffi/closure.cr
+++ b/src/compiler/crystal/ffi/closure.cr
@@ -3,7 +3,7 @@ require "./ffi"
 module Crystal::FFI
   struct Closure
     def initialize(call_interface : CallInterface, closure_fun : LibFFI::ClosureFun, user_data : Void*)
-      @closure = LibFFI.closure_alloc(sizeof(LibFFI::Closure), out @code)
+      @closure = LibFFI.closure_alloc(LibFFI::SIZEOF_CLOSURE, out @code)
       unless @closure
         raise "Error on LibFFI.closure_alloc"
       end

--- a/src/compiler/crystal/ffi/lib_ffi.cr
+++ b/src/compiler/crystal/ffi/lib_ffi.cr
@@ -121,30 +121,23 @@ module Crystal
     $ffi_type_double : Type
     $ffi_type_pointer : Type
 
-    # TODO: this is 12 for non-x
+    # We're treating ffi_closure as an opaque type because we do not need to interact
+    # with it directly and only pass a pointer around. We only need to allocate
+    # the memory, but the memory size differs based on target and ABI version (https://github.com/libffi/libffi/pull/540)
+    # We define the maximum possible size for each target and allocate that amount
+    # of memory, even if less would suffice. Overallocating a couple of bytes should
+    # not cause any issues
+    # https://github.com/crystal-lang/crystal/pull/12192#issuecomment-1173993292
+    # https://github.com/libffi/libffi/blob/ddc6764386b29449d941b2b18d000f2987a9d848/doc/libffi.texi#L815
+    type Closure = Void
 
-    {% if compare_versions(`hash pkg-config 2> /dev/null && pkg-config --modversion libffi`.chomp, "3.4.0") >= 0 %}
-      {% if flag?(:bits64) %}
-        FFI_TRAMPOLINE_SIZE = 32
-      {% else %}
-        FFI_TRAMPOLINE_SIZE = 16
-      {% end %}
+    {% if flag?(:bits64) %}
+      SIZEOF_CLOSURE = 56
     {% else %}
-      {% if flag?(:bits64) %}
-        FFI_TRAMPOLINE_SIZE = 24
-      {% else %}
-        FFI_TRAMPOLINE_SIZE = 12
-      {% end %}
+      SIZEOF_CLOSURE = 40
     {% end %}
 
     alias ClosureFun = Cif*, Void*, Void**, Void* -> Void
-
-    struct Closure
-      tramp : LibC::Char[FFI_TRAMPOLINE_SIZE]
-      cif : Cif*
-      fun : ClosureFun
-      user_data : Void*
-    end
 
     fun prep_cif = ffi_prep_cif(
       cif : Cif*,

--- a/src/compiler/crystal/ffi/lib_ffi.cr
+++ b/src/compiler/crystal/ffi/lib_ffi.cr
@@ -123,10 +123,18 @@ module Crystal
 
     # TODO: this is 12 for non-x
 
-    {% if flag?(:bits64) %}
-      FFI_TRAMPOLINE_SIZE = 24
+    {% if compare_versions(`hash pkg-config 2> /dev/null && pkg-config --modversion libffi`.chomp, "3.4.0") >= 0 %}
+      {% if flag?(:bits64) %}
+        FFI_TRAMPOLINE_SIZE = 32
+      {% else %}
+        FFI_TRAMPOLINE_SIZE = 16
+      {% end %}
     {% else %}
-      FFI_TRAMPOLINE_SIZE = 12
+      {% if flag?(:bits64) %}
+        FFI_TRAMPOLINE_SIZE = 24
+      {% else %}
+        FFI_TRAMPOLINE_SIZE = 12
+      {% end %}
     {% end %}
 
     alias ClosureFun = Cif*, Void*, Void**, Void* -> Void

